File: team-event-management-services/functions/src/domain/entities/Event.ts
/**
 * Represents the Event aggregate root in the domain.
 * An Event is a scheduled activity that can be assigned to users or teams.
 */
export class Event {
  /**
   * The unique identifier for the event.
   */
  public readonly id: string;

  /**
   * The identifier of the tenant this event belongs to.
   */
  public readonly tenantId: string;

  /**
   * The user ID of the person who created the event.
   */
  public readonly createdByUserId: string;

  public title: string;
  public description?: string;
  public startTime: Date;
  public endTime: Date;
  public assignedUserIds: Set<string>;
  public assignedTeamIds: Set<string>;
  public recurrenceRule?: string;

  /**
   * @param id The unique identifier for the event.
   * @param tenantId The identifier of the tenant.
   * @param createdByUserId The user ID of the event creator.
   * @param title The title of the event.
   * @param startTime The start date and time of the event.
   * @param endTime The end date and time of the event.
   * @param description An optional description of the event.
   * @param assignedUserIds An array of user IDs directly assigned to the event.
   * @param assignedTeamIds An array of team IDs assigned to the event.
   * @param recurrenceRule An optional RRULE string for recurring events.
   */
  constructor(
    id: string,
    tenantId: string,
    createdByUserId: string,
    title: string,
    startTime: Date,
    endTime: Date,
    description?: string,
    assignedUserIds: string[] = [],
    assignedTeamIds: string[] = [],
    recurrenceRule?: string,
  ) {
    if (!id || !tenantId || !createdByUserId || !title || !startTime || !endTime) {
      throw new Error("Event entity requires id, tenantId, createdByUserId, title, startTime, and endTime.");
    }

    if (endTime <= startTime) {
      throw new Error("Event end time must be after the start time.");
    }

    this.id = id;
    this.tenantId = tenantId;
    this.createdByUserId = createdByUserId;
    this.title = title;
    this.description = description;
    this.startTime = startTime;
    this.endTime = endTime;
    this.assignedUserIds = new Set(assignedUserIds);
    this.assignedTeamIds = new Set(assignedTeamIds);
    this.recurrenceRule = recurrenceRule;
  }

  /**
   * Checks if the event is a recurring event.
   * @returns True if the event has a recurrence rule, false otherwise.
   */
  public isRecurring(): boolean {
    return !!this.recurrenceRule && this.recurrenceRule.length > 0;
  }

  /**
   * Updates the details of the event.
   * @param details An object containing the event details to update.
   */
  public updateDetails(details: {
    title?: string;
    description?: string;
    startTime?: Date;
    endTime?: Date;
  }): void {
    const newStartTime = details.startTime ?? this.startTime;
    const newEndTime = details.endTime ?? this.endTime;

    if (newEndTime <= newStartTime) {
      throw new Error("Event end time must be after the start time.");
    }

    this.title = details.title ?? this.title;
    this.description = details.description ?? this.description;
    this.startTime = newStartTime;
    this.endTime = newEndTime;
  }

  /**
   * Assigns the event to a specific user.
   * @param userId The ID of the user to assign.
   */
  public assignToUser(userId: string): void {
    this.assignedUserIds.add(userId);
  }

  /**
   * Unassigns a user from the event.
   * @param userId The ID of the user to unassign.
   */
  public unassignFromUser(userId: string): void {
    this.assignedUserIds.delete(userId);
  }

  /**
   * Assigns the event to a specific team.
   * @param teamId The ID of the team to assign.
   */
  public assignToTeam(teamId: string): void {
    this.assignedTeamIds.add(teamId);
  }

  /**
   * Unassigns a team from the event.
   * @param teamId The ID of the team to unassign.
   */
  public unassignFromTeam(teamId: string): void {
    this.assignedTeamIds.delete(teamId);
  }
}

File: client-data-access/lib/src/features/event/domain/entities/event.dart
import 'package:equatable/equatable.dart';

/// Enum representing the frequency of a recurring event.
enum RecurrenceFrequency {
  daily,
  weekly,
  monthly,
}

/// A value object representing the recurrence rule for an event.
class RecurrenceRule extends Equatable {
  /// The frequency of the recurrence.
  final RecurrenceFrequency frequency;

  /// The interval of recurrence (e.g., every 2 weeks). Defaults to 1.
  final int interval;

  /// The date until which the event recurs (inclusive).
  final DateTime until;

  /// For weekly recurrence, the days of the week (1=Monday, 7=Sunday).
  final List<int>? byDay;

  const RecurrenceRule({
    required this.frequency,
    this.interval = 1,
    required this.until,
    this.byDay,
  });

  @override
  List<Object?> get props => [frequency, interval, until, byDay];
}


/// Represents a scheduled event within a tenant.
/// This is a pure domain entity.
class Event extends Equatable {
  /// The unique identifier for the event.
  final String id;

  /// The ID of the tenant this event belongs to.
  final String tenantId;

  /// The title of the event.
  final String title;

  /// A detailed description of the event.
  final String? description;

  /// The start time of the event.
  final DateTime startTime;

  /// The end time of the event.
  final DateTime endTime;

  /// The ID of the user who created the event.
  final String createdByUserId;

  /// A list of user IDs directly assigned to this event.
  final List<String> assignedUserIds;

  /// A list of team IDs assigned to this event.
  final List<String> assignedTeamIds;
  
  /// The recurrence rule for this event, if it's a recurring event.
  final RecurrenceRule? recurrenceRule;
  
  /// If this is an instance of a recurring event, this ID links it to the master event.
  final String? recurrenceId;

  const Event({
    required this.id,
    required this.tenantId,
    required this.title,
    this.description,
    required this.startTime,
    required this.endTime,
    required this.createdByUserId,
    this.assignedUserIds = const [],
    this.assignedTeamIds = const [],
    this.recurrenceRule,
    this.recurrenceId,
  });
  
  /// Returns true if this is a recurring event master.
  bool get isRecurring => recurrenceRule != null;

  /// Calculates the duration of the event.
  Duration get duration => endTime.difference(startTime);

  @override
  List<Object?> get props => [
        id,
        tenantId,
        title,
        description,
        startTime,
        endTime,
        createdByUserId,
        assignedUserIds,
        assignedTeamIds,
        recurrenceRule,
        recurrenceId,
      ];
}